UnitTests/ControllerTests/EPubsControllerTest.cs

```diff
@@ -0,0 +1,216 @@
+                    SourceType = ResourceType.Course
+                },
+                new EPub
+                {
+                    Title = "example3",
+                    Language = "en-US",
+                    Filename = "filename_example",
+                    Author = "author_example",
+                    Publisher = "publisher_example",
+                    SourceId = "mediaId_example",
+                    SourceType = ResourceType.Media
+                }
+            };
+
+            foreach (var ePub in ePubs)
+            {
+                var postResult = await _controller.PostEPub(ePub);
+                var createdResult = postResult.Result as CreatedAtActionResult;
+
+                Assert.NotNull(createdResult);
+            }
+
+            var getResult = await _controller.GetEPubsBySource(ResourceType.Media.ToString(), ePubs[0].SourceId);
+            List<EPub> ePubsBySource = getResult.Value.ToList();
+
+            Assert.Equal(2, ePubsBySource.Count);
+            Assert.Equal(ePubs[0], ePubsBySource[0]);
+            Assert.Equal(ePubs[2], ePubsBySource[1]);
+        }
+
+        [Fact]
+        public async Task Post_Invalid_and_Valid_EPubs()
+        {
+            var ePub = new EPub
+            {
+                Title = "example0",
+                Language = "en-US"
+            };
+
+            var postResult = await _controller.PostEPub(ePub);
+            var badRequestResult = postResult.Result as BadRequestObjectResult;
+
+            Assert.NotNull(badRequestResult);
+
+            ePub.Filename = "filename_example";
+            ePub.Author = "author_example";
+
+            postResult = await _controller.PostEPub(ePub);
+            badRequestResult = postResult.Result as BadRequestObjectResult;
+
+            Assert.NotNull(badRequestResult);
+
+            ePub.Publisher = "publisher_example";
+            ePub.SourceId = "mediaId_example";
+            ePub.SourceType = ResourceType.Media;
+
+            postResult = await _controller.PostEPub(ePub);
+            var createdResult = postResult.Result as CreatedAtActionResult;
+
+            Assert.NotNull(createdResult);
+        }
+
+        [Fact]
+        public async Task Put_Invalid_EPubs()
+        {
+            var ePub = new EPub
+            {
+                Title = "example0",
+                Language = "en-US",
+                Filename = "filename_example",
+                Author = "author_example",
+                Publisher = "publisher_example",
+                SourceId = "mediaId_example",
+                SourceType = ResourceType.Media
+            };
+
+            var postResult = await _controller.PostEPub(ePub);
+            var createdResult = postResult.Result as CreatedAtActionResult;
+
+            Assert.NotNull(createdResult);
+
+            ePub.Title = string.Empty;
+            ePub.Filename = string.Empty;
+
+            var putResult = await _controller.PutEPub(ePub.Id, ePub);
```

Comment:
A bit off topic, are these constraints visible somewhere in `/swag`? Or is the only way to try it and see the 400?

https://github.com/classtranscribe/WebAPI/blob/50cb4b943d881f92e8adecdf298da04b2ddac7d6/ClassTranscribeServer/Controllers/EPubsController.cs#L164-L172

Reply:
The most authoratitive (and fastest check) is the  controller source code e.g.
https://github.com/classtranscribe/WebAPI/blob/50cb4b943d881f92e8adecdf298da04b2ddac7d6/ClassTranscribeServer/Controllers/WatchHistoriesController.cs#L52

Also the method decoration (e.g. see the AdminController
[Authorize(Roles = Globals.ROLE_ADMIN)]
will cause notauthorized if you don't have that role
e.g.
https://github.com/classtranscribe/WebAPI/blob/50cb4b943d881f92e8adecdf298da04b2ddac7d6/ClassTranscribeServer/Controllers/AdminController.cs#L159

I believe that magic is set up here
https://github.com/classtranscribe/WebAPI/blob/50cb4b943d881f92e8adecdf298da04b2ddac7d6/ClassTranscribeServer/Startup.cs#L94
and at the end of the same file...
            app.UseAuthorization();
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapDefaultControllerRoute().RequireAuthorization();
            });